### PR TITLE
Add SOA entropy regularization methods to tf/NPO

### PR DESCRIPTION
**Pull request by CatherineSue**
* Add max entropy method to tf/NPO.
* Fix the neg_log_likeli entropy estimation in tf/NPO.
* Add documentation of sane entropy regularization default.
* Replace Enum params with strings.

### CODE REVIEW

**Comment by CatherineSue** [comment]
### Entropy Experiments
I've experimented the `max` and `regularized` entropy methods with both GaussianMLPPolicy and GaussianLSTMPolicy.
Here are some sane defaults:
`entropy_method='max'`, `center_adv=False`, `stop_gradient=True`
`entropy_method='regularized'`, `stop_gradient=False`, `use_neg_logli_entropy=False`

The following curve shows the entropy of GaussianMLPPolicy and GaussianLSTMPolicy. Curves that have increasing entropy have `policy_ent_coeff=1.0`. Curves that have decreasing entropy have `policy_ent_coeff=-1.0`. Note that it takes longer for GaussianLSTMPolicy to train, so I stopped their training before the `max_iter` is reached.

Screen Shot 2019-04-23 at 3 56 04 PM
Screen Shot 2019-04-23 at 4 02 17 PM

File: garage/tf/algos/npo.py [hunk 1]
```diff
@@ -33,14 +33,33 @@ class NPO(BatchPolopt):
     """
     Natural Policy Gradient Optimization.
 
-    Attributes:
-        name(str): The name of the algorithm.
+    Args:
+        pg_loss(enum): The type of loss functions to use.
```
Comment (ryanjulian) [inline garage/tf/algos/npo.py]: you can call out the specific enum type (PGLoss) rather than just a generic enum

Note that we have issue #435 open to remove this enum anyway. Perhaps now is a good time to just replace it with the string?

File: garage/tf/algos/npo.py [hunk 2]
```diff
@@ -33,14 +33,33 @@ class NPO(BatchPolopt):
...
+        optimizer_args(dict): The arguments of the optimizer.
+        name(str): The name of the algorithm.
         policy_ent_coeff(float): The coefficient of the policy entropy.
-        optimizer(float): The optimizer of the algorithm.
+        entropy_method(str): A string from: 'max', 'regularized'. The type
+            of entropy method to use. 'max' adds the dense entropy to the
+            reward for each time step. 'regularized' adds the mean entropy
+            to the surrogate objective. See https://arxiv.org/abs/1805.00909
+            for more details.
+            !!NOTE!!: make sure when setting entropy_method to 'max', set
```
Comment (ryanjulian) [inline garage/tf/algos/npo.py]: Please put this as a note outside the argument list.

```python
"""
Args:
    foo (str): Stuff

Notes
    Notes about some stuff
"""
```

See example here https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html

File: garage/tf/algos/npo.py [hunk 3]
```diff
@@ -33,14 +33,33 @@ class NPO(BatchPolopt):
     """
     Natural Policy Gradient Optimization.
 
-    Attributes:
-        name(str): The name of the algorithm.
+    Args:
```
Comment (ryanjulian) [inline garage/tf/algos/npo.py]: please also note which params are optional

File: garage/tf/algos/npo.py [hunk 4]
```diff
@@ -33,14 +33,33 @@ class NPO(BatchPolopt):
...
+        name(str): The name of the algorithm.
         policy_ent_coeff(float): The coefficient of the policy entropy.
-        optimizer(float): The optimizer of the algorithm.
+        entropy_method(str): A string from: 'max', 'regularized'. The type
+            of entropy method to use. 'max' adds the dense entropy to the
+            reward for each time step. 'regularized' adds the mean entropy
+            to the surrogate objective. See https://arxiv.org/abs/1805.00909
+            for more details.
+            !!NOTE!!: make sure when setting entropy_method to 'max', set
+            `center_adv` to False and turn off policy gradient. `center_adv`
```
Comment (ryanjulian) [inline garage/tf/algos/npo.py]: long comments like this belong elsewhere in the docstring. argument descriptions should be short

File: garage/tf/algos/npo.py [hunk 5]
```diff
@@ -50,31 +69,39 @@ def __init__(self,
...
             optimizer = LbfgsOptimizer
 
+        if entropy_method == 'max':
+            self._maximum_entropy = True
+            self._entropy_regularzied = False
+        elif entropy_method == 'regularized':
+            self._maximum_entropy = False
+            self._entropy_regularzied = True
+        else:
+            raise NotImplementedError('Unknown entropy_method')
```
Comment (ryanjulian) [inline garage/tf/algos/npo.py]: this is a ValueError

File: garage/tf/algos/npo.py [hunk 6]
```diff
@@ -33,14 +33,33 @@ class NPO(BatchPolopt):
...
 
-    Attributes:
-        name(str): The name of the algorithm.
+    Args:
+        pg_loss(enum): The type of loss functions to use.
         lr_clip_range(float): The limit on the likelihood ratio between
             policies, as in PPO.
         max_kl_step(float): The maximum KL divergence between old and new
             policies, as in TRPO.
+        optimizer(object): The optimizer of the algorithm.
```
Comment (ryanjulian) [inline garage/tf/algos/npo.py]: is there an optimizer base class we can put here instead?

**Comment by ryanjulian** [comment]
ref #628

File: garage/tf/algos/npo.py [hunk 7]
```diff
@@ -33,14 +33,33 @@ class NPO(BatchPolopt):
...
+        optimizer_args(dict): The arguments of the optimizer.
+        name(str): The name of the algorithm.
         policy_ent_coeff(float): The coefficient of the policy entropy.
-        optimizer(float): The optimizer of the algorithm.
+        entropy_method(str): A string from: 'max', 'regularized'. The type
```
Comment (naeioi) [inline garage/tf/algos/npo.py]: Is adding entropy penalty to loss or reward enforced (instead of an option)?

File: garage/tf/algos/npo.py (hunk 6 above)
Comment (CatherineSue) [reply garage/tf/algos/npo.py]: No I don't think there is one. We don't have optimizer base class in garage/tf.

Comment (ryanjulian) [reply garage/tf/algos/npo.py]: Ok it's fine. we can just say (object) and note the optimizer package in the description

File: garage/tf/algos/npo.py (hunk 7 above)
Comment (ryanjulian) [reply garage/tf/algos/npo.py]: it's an option. you can also train with no entropy regularization.

also note that entropy is a bonus, not a penalty (we want more of it)

Comment (CatherineSue) [reply garage/tf/algos/npo.py]: should we add `None` check to `entropy_method`? my assumption is users can set `policy_ent_coeff=0.0` if they don't want to use entropy regularization

File: garage/tf/algos/npo.py (hunk 5 above)
Comment (CatherineSue) [reply garage/tf/algos/npo.py]: there is no third way to configure the entropy bonus?

**Comment by ryanjulian** [comment]
ref #435

File: garage/tf/algos/npo.py (hunk 7 above)
Comment (ryanjulian) [reply garage/tf/algos/npo.py]: hmm. you should do what you can to prevent the user from providing mutually-incompatible arguments.

e.g. a non-zero `policy_ent_coeff` when `entropy_method` is not `'no_entropy'` should be an error. Note that I recommend that the entropy method for none be `'no_entropy'` rather than the Python symbol `None`, because `None` represents that the user failed to make a choice or provide a value, *not* that the user has *chosen* to not use entropy.

Comment (ryanjulian) [reply garage/tf/algos/npo.py]: please add similar checks for other mutually-incompatible arguments, as you have noted in the comments.

only disallow arguments which make no sense together (e.g. guaranteed to fail or logically inconsistent). don't disallow argument combinations which seem useless or unlikely but not harmful.

**Comment by ryanjulian** [comment]
overall this still LGTM, but please see my note on validating arguments. 

it would also be nice if you could push up the coverage on NPO (it stands at 86%). Based on CodeCov, it seems like a couple of the entropy cases might not be fully-covered: https://codecov.io/gh/rlworkgroup/garage/compare/3c49440b2692045c8c9b0db86a1e550d1bca6e84...4d802f847d9cc8ce014d2d2fb975045f40b73e23/diff

Perhaps this is because some combinations of arguments don't make sense, though?

**Comment by CatherineSue** [comment]
This PR is ready for another review.

File: garage/tf/algos/npo.py [hunk 8]
```diff
@@ -22,59 +20,101 @@
...
+        use_neg_logli_entropy (bool, optional): Whether to estimate the
+            entropy as the negative log likelihood of the action.
+        use_softplus_entropy (bool, optional): Whether to estimate the softmax
+            distribution of the entropy to prevent the entropy from being
+            negative.
+        stop_entropy_gradient (bool, optional): Whether to stop the entropy
+            gradient.
+
+    Note:
+        sane defaults for entropy configuration:
```
Comment (ryanjulian) [inline garage/tf/algos/npo.py]: can you add some more explicit info on the combinations of entropy arguments which are valid? even i find it a little confusing.

File: garage/tf/algos/npo.py (hunk 3 above)
Comment (ryanjulian) [reply garage/tf/algos/npo.py]: on second thought, the documentation generator makes it clear what's optional, and these notations are just noise. please remove.

i apologize for changing my mind.

**Comment by ryanjulian** [comment]
Are these uncovered branches in CodeCov yellow because they are not covered, or because the missed branches are not valid combinations of configuration?

https://codecov.io/gh/rlworkgroup/garage/compare/8e24d281b3f299663528034191f8f54934d5c6e7...f2ac9b6e0fe1b9673d6d6c8d04cb457bbdce8144/diff#D2-x423

File: garage/tf/algos/npo.py (hunk 8 above)
Comment (CatherineSue) [reply garage/tf/algos/npo.py]: Could you tell me what information you think is missing here?

**Comment by CatherineSue** [comment]
I didn’t test npo with turning on use_softplus_entropy because I didn’t see any negative entropy in experiments. I’m not sure if it is an invalid configuration.
Another yellow line in ppo is because the default arguments of FirstOrderOptimizer has max_epochs=1000. It takes too long to test it. We can change the default max_epochs to a smaller number, like 10 or 20 and then test it.

**Comment by ryanjulian** [comment]
This seems fine to me -- I was just curious.

max_epochs (which should actually be max_minibatches, right?) is pretty big. Feel free to make it more sane if it doesn't change much.

File: garage/tf/algos/npo.py (hunk 3 above)
Comment (CatherineSue) [reply garage/tf/algos/npo.py]: Which documentation generator can make it clear about optional parameters? I tried with pydoc and pdoc but they don’t specify optional parameters.

Comment (ryanjulian) [reply garage/tf/algos/npo.py]: For instance, in akro the module documentation looks like this:
https://akro.readthedocs.io/en/latest/_apidoc/akro.html#module-akro.box

You can see the function signature with kwargs, so it's obvious the kwargs are optional and the args are not.

File: garage/tf/algos/npo.py (hunk 8 above)
Comment (ryanjulian) [reply garage/tf/algos/npo.py]: sorry, this seems sufficient to me.

File: tests/garage/tf/algos/test_npo.py [hunk 9]
```diff
@@ -43,9 +43,9 @@ def test_npo_pendulum(self):
 
             env.close()
 
-    def test_npo_unknown_pg_loss(self):
+    def test_npo_with_unknown_pg_loss(self):
         """Test NPO with unkown policy gradient loss."""
-        env = TfEnv(normalize(gym.make("InvertedDoublePendulum-v2")))
+        env = TfEnv(normalize(gym.make('InvertedDoublePendulum-v2')))
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply tests/garage/tf/algos/test_npo.py]: Done.

File: tests/garage/tf/algos/test_trpo.py [hunk 10]
```diff
@@ -68,8 +68,37 @@ def test_trpo_unknown_kl_constraint(self):
...
+                max_path_length=100,
+                discount=0.99,
+                gae_lambda=0.98,
+                policy_ent_coeff=0.0,
+                kl_constraint='soft')
+            runner.setup(algo, env)
+            last_avg_ret = runner.train(n_epochs=10, batch_size=2048)
+            assert last_avg_ret > 50
+
+            env.close()
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply tests/garage/tf/algos/test_trpo.py]: Done.

File: garage/tf/algos/npo.py [hunk 11]
```diff
@@ -355,7 +399,8 @@ def _build_policy_loss(self, i):
                 else:
                     raise NotImplementedError('Unknown PGLoss')
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply garage/tf/algos/npo.py]: Moved the check to the constructor. I used `in` to check if `pg_loss` is valid because in the constructor we only need to check if an unknown value is passed.

File: garage/tf/algos/npo.py [hunk 12]
```diff
@@ -22,59 +20,101 @@
...
         self._use_neg_logli_entropy = use_neg_logli_entropy
         self._stop_entropy_gradient = stop_entropy_gradient
-
         self._pg_loss = pg_loss
         if optimizer is None:
             if optimizer_args is None:
                 optimizer_args = dict()
             optimizer = LbfgsOptimizer
 
+        if entropy_method == 'max':
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply garage/tf/algos/npo.py]: Moved.

File: tests/garage/tf/algos/test_trpo_with_model.py [hunk 13]
```diff
@@ -36,6 +36,6 @@ def test_trpo_recurrent_cartpole(self):
 
             runner.setup(algo, env)
             last_avg_ret = runner.train(n_epochs=10, batch_size=2048)
-            assert last_avg_ret > 85
+            assert last_avg_ret > 75
```
Comment (CatherineSue) [inline tests/garage/tf/algos/test_trpo_with_model.py]: 85 seems still to be flaky as it didn't pass the ci check. I changed it to 75.

File: src/garage/tf/algos/ppo.py [hunk 14]
```diff
@@ -61,29 +71,31 @@ def __init__(self,
                  use_softplus_entropy=False,
                  use_neg_logli_entropy=False,
                  stop_entropy_gradient=False,
+                 entropy_method='no_entropy',
                  name='PPO'):
         if optimizer is None:
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply src/garage/tf/algos/ppo.py]: Changed. The reason we didn't add it to default is that in some algorithms the default optimizer can't be decided in `__init__`. It is rare though.

File: src/garage/tf/algos/tnpg.py [hunk 15]
```diff
@@ -65,21 +65,14 @@ def __init__(self,
...
                  stop_entropy_gradient=False,
                  entropy_method='no_entropy',
                  name='TNPG'):
-        if optimizer is None:
-            optimizer = ConjugateGradientOptimizer
-            default_args = dict(max_backtracks=1)
-            if optimizer_args is None:
-                optimizer_args = default_args
-            else:
-                optimizer_args = dict(default_args, **optimizer_args)
```
Comment (zequnyu) [inline src/garage/tf/algos/tnpg.py]: According to the logic here we want to keep `max_backtracks=1 ` for the `optimizer_args ` even if the user passes their own args. I'm not sure if it's necessary to be this value for this particular algorithm. (I traced back to `ConjugateGradientOptimizer` where the default `max_backtracks` is 15.)

Comment (CatherineSue) [reply src/garage/tf/algos/tnpg.py]: It is probably necessary. I think setting `max_backtracks` too large will probably affect the training performance of TNPG since it doesn't use the backtrack line search to make sure that the trust region is not broken.

Comment (zequnyu) [reply src/garage/tf/algos/tnpg.py]: The danger here is that if the user passes some args other than `max_backtracks` it will override the default value. But the previous one (using `default_args`) also has a flaw: `max_backtracks` can not be customised otherwise it raises error for duplicate attribute values.

Comment (CatherineSue) [reply src/garage/tf/algos/tnpg.py]: No. The previous one won't raise duplicate attribute because python will overwrite the value of an existing key.

Comment (CatherineSue) [reply src/garage/tf/algos/tnpg.py]: I don't think the previous one is good because the values in `optimizer_args` is likely to overwrite the values in `default_args`.

Comment (CatherineSue) [reply src/garage/tf/algos/tnpg.py]: I prefer the current because I don't think we should constrain the `max_backtracks=1` or `max_epochs=1` in algorithms. It is unlikely that the algorithm won't be working when `max_epochs=2`. 
If we need fix this, I think set sane defaults in optimizer is better? For instance, the default `max_epochs` in `FirstOrderOptimizer` is 100, that is not a sane default. Even for a small and simple environment, running each iteration in 100 epochs is going to take a lot of time.

Comment (CatherineSue) [reply src/garage/tf/algos/tnpg.py]: Sorry, on second thought, I don't really have a preference. I think each has its drawbacks. If it is ok with you, I would like to revert this change (both `optimizer` and `optimizer_args`). I think it is not very important for this PR.

Comment (zequnyu) [reply src/garage/tf/algos/tnpg.py]: Sure. Ok with me.
